jlm/tooling/CommandLine.hpp

```diff
@@ -160,7 +160,7 @@ class JlmOptCommandLineOptions final : public CommandLineOptions
   static const char *
   ToCommandLineArgument(OutputFormat outputFormat);
 
-  [[nodiscard]] std::unique_ptr<llvm::optimization>
+  [[nodiscard]] llvm::optimization *
```

Comment:
If this method uses a static map internally, I think the method itself should also be static, to avoid accidentally using state

Reply:
I do not fully see how to do this (without some major refactoring). This method calls `CreateOptimization()`, which reads internal state. Thus, I cannot just make this method static as the `CreateOptimization()` method needs the state.

Reply:
@phate
If the created optimizations rely on state, that is an argument against including them in a global map, and not per-class map. The next instance of `JlmCommandLineOptions` risks returning optimizations initialized with the wrong data.

That being said, almost all the optimizations are "stateless" when they are not running, so it is mostly not an issue (unless we start running multiple JlmOpt commands in parallel in the same binary any time soon).

The fact that some optimizations rely on state / have a reference to a per `JlmCommandLineOptions` configuration, makes me prefer a solution where it is possible to create separate instances of optimizations (if only those with state). This would also allow the Tree dumper to have its own counter, for instance. Making the map a field instead of a global variable would achieve that, at the expense of making multiple instances of some singletons in very specific circumstances.

Reply:
@haved I thought about this a bit more and realized that the JlmCommandLineOption class is the wrong to place to have these optimizations in the first place. I moved the entire optimization creation into the JlmOptCommand class now. There are no static variables used any longer. It is much better now (in my opinion). Thanks for being persistent!